File: examples/src/fireworks/java/Firework.java
import org.joml.Vector2f;
import org.joml.Vector4f;

import java.util.HashSet;
import java.util.Set;

import static io.github.bernhardfritz.p8g.Sketch.noStroke;

public class Firework {

    private Particle rocket;
    boolean exploded;
    private Set<Particle> particles;

    public Firework(Particle rocket) {
        this.rocket = rocket;
        this.exploded = false;
        this.particles = new HashSet<>();
    }

    public void applyForce(Vector2f force) {
        if (!exploded) {
            rocket.applyForce(force);
        } else {
            for (Particle particle : particles) {
                particle.applyForce(force);
            }
        }
    }

    public void update() {
        if (!exploded) {
            rocket.update();
            if (rocket.getVelocity().get(1) >= 0.f) {
                explode();
            }
        } else {
            for (Particle particle : particles) {
                particle.getVelocity().mul(0.95f);
                particle.update();
            }
            particles.removeIf(Particle::isDead);
        }
    }

    public void explode() {
        Vector4f color = new Vector4f(Random.randf(0.f, 255.f), Random.randf(0.f, 255.f), Random.randf(0.f, 255.f), 255.f);
        for (int i = 0; i < 200; i++) {
            float angle = Random.randf(0.f, 2.f * (float) Math.PI);
            float magnitude = Random.randf(2.f, 10.f);
            Particle particle = new Particle(
                    new Vector2f(rocket.getPosition()),
                    Random.randf(0.7f, 1.f),
                    new Vector2f(magnitude * (float) Math.cos(angle), magnitude * (float) Math.sin(angle)),
                    (int) Random.randf(90.f, 110.f),
                    color
            );
            particles.add(particle);
        }
        exploded = true;
    }

    public void draw() {
        noStroke();
        if (!exploded) {
            rocket.draw();
        } else {
            for (Particle particle : particles) {
                particle.draw();
            }
        }
    }

    public boolean isDone() {
        return exploded && particles.isEmpty();
    }

}

File: examples/src/fireworks/c/firework.c
#include "firework.h"

#include "stb_ds.h"
#define USING_NAMESPACE_P8G
#include "p8g.h"

void firework_apply_force(firework_t* firework, vec2 force) {
    if (firework->exploded == 0) {
        particle_t* rocket = &firework->rocket;
        particle_apply_force(rocket, force);
    } else {
        for (int i = 0; i < arrlen(firework->particles); i++) {
            particle_t* particle = &firework->particles[i];
            particle_apply_force(particle, force);
        }
    }
}

void firework_update(firework_t* firework) {
    if (firework->exploded == 0) {
        particle_t* rocket = &firework->rocket;
        particle_update(rocket);
        if (rocket->velocity[1] >= 0.f) {
            firework_explode(firework);
        }
    } else {
        for (int i = 0; i < arrlen(firework->particles); i++) {
            particle_t* particle = &firework->particles[i];
            vec2_scale(particle->velocity, particle->velocity, 0.95f);
            particle_update(particle);
        }
        for (int i = arrlen(firework->particles) - 1; i >= 0; i--) {
            particle_t* particle = &firework->particles[i];
            if (particle_is_dead(particle)) {
                arrdel(firework->particles, i);
            }
        }
    }
}

void firework_explode(firework_t* firework) {
    vec4 color = { random(255.f), random(255.f), random(255.f), 255.f };
    for (int i = 0; i < 200; i++) {
        float angle = random(2.f * M_PI);
        float magnitude = random(2.f, 10.f);
        particle_t particle = {
            .position = { firework->rocket.position[0], firework->rocket.position[1] },
            .mass = random(0.7f, 1.f),
            .velocity = { magnitude * cosf(angle), magnitude * sinf(angle) },
            .lifespan = random(90.f, 110.f),
            .color = { color[0], color[1], color[2], color[3] },
        };
        arrpush(firework->particles, particle);
    }
    firework->exploded = 1;
}

void firework_draw(firework_t* firework) {
    noStroke();
    if (firework->exploded == 0) {
        particle_t* rocket = &firework->rocket;
        particle_draw(rocket);
    } else {
        for (int i = 0; i < arrlen(firework->particles); i++) {
            particle_t* particle = &firework->particles[i];
            particle_draw(particle);
        }
    }
}

int firework_is_done(firework_t* firework) {
    return firework->exploded == 1 && arrlen(firework->particles) == 0
        ? 1
        : 0;
}

void firework_destroy(firework_t* firework) {
    arrfree(firework->particles);
}